### ISSUE TITLE
Error in Realtime Ingestion Marks Segment as Offline in Ideal State

### ISSUE THREAD

**Issue by ankitsultana** [apache/pinot#11314]
We saw an issue recently where due to a intermittent DNS issue, there was an error in KafkaConsumer init call which led to a failure in the PartitionConsumer thread. This happened in 1 of the two replicas of the segment.

After this happened, the server sent a "segmentStoppedConsuming" message to the Controller, and the controller marked the segment as OFFLINE in the Ideal State in that replica.

We had to wait for the segment to commit for this to resolve (we could have used forceCommit as well but I haven't run that before so didn't want to take any chances).

Is there a better way to fix the segment in this case? Segment reset or restarting the server doesn't help because the Ideal State itself says that the segment is offline.

**Comment by Jackie-Jiang** [comment]
@mcvsubbu Do you remember why we mark IS as OFFLINE? Is that just a workaround because we don't know how to mark EV as ERROR?

**Comment by mcvsubbu** [comment]
Marking it offline gave us the flow correctly in terms of dropping all resources in the server.

@ankitsultana  the way to recover is to run the realtime segment validation job on the controller after the error situation is corrected.

**Comment by mcvsubbu** [comment]
cc: @sajjad-moradi

**Comment by ankitsultana** [comment]
> @ankitsultana the way to recover is to run the realtime segment validation job on the controller after the error situation is corrected.

I think that only helps if both replicas are offline. If only one replica is lost then even the segment validation job doesn't do anything. In our case we have the segment validation job run every 1 hour, and 1 of the replicas remained offline for several hours.

**Comment by mcvsubbu** [comment]
Correct, I realized after posting that you may have had a problem with one of the replicas.

So, the thought was as follows:

In general, if there are consumption issues from a stream, it is likely that all replicas face issues communicating with the stream. So the likely case is that all replicas go down eventually, perhaps within short time of each other.

Further, if only a subset of replicas go down, there is still at least one replica alive to complete the segment. Once the segment is complete, the transition from OFFLINE to ONLINE would arrive, and everything would return back to normal.The thought here is that the other replicas are enough to serve queries. After all , only one segment will be routed to some other replica. The server in question can continue to serve other segments of the table.

Lastly, let us say we try to modify the state back to CONSUMING in idealstate. We will need to worry about race conditions where the segment may be in completing stage. Also, the host may have to feverishly catch up to the data only to find that some other replica is at the point of completing the segment.

For these reasons, we decided to fix the situation only if all replicas are OFFLINE.

**Comment by ankitsultana** [comment]
@mcvsubbu : thanks for confirming, this makes sense.

Do you think it would make sense to add a periodic task in Controller that can detect this scenario and issue a force commit on the table to fix this? It could be an optional task that would be disabled by default.

In many of our internal use-cases the segments take quite a while to commit, and we have to do that to ensure the number of segments aren't too high. So when this issue happens, we are left with only 1 copy of a segment for several hours which leads to reliability dips particularly during rollouts.

I haven't looked into force commit enough yet so lmk if this could lead to another set of issues.

cc: @Jackie-Jiang as well for input.

**Comment by mcvsubbu** [comment]
We have a realtime segment validation periodic job that detects all-replicas-offline condition. Maybe it can be augmented to detect the condition you describe, and issue a force-commit (if confgured to do so). I would suggest to verify that the condition existed at least for X minutes before triggering forcecommit.

Alternatvely, another job may be easier, and folks can configure that in or out, but the read of zk will be done again in the second job. I forget whether we list all segments and sort them, but something to look at.

**Comment by Jackie-Jiang** [comment]
I'd suggest reusing the existing periodic task with an extra flag to handle the case when partial segments are offline. The problem with 2 separate periodic tasks is that they might intervene with each other, and both of them are trying to do the same thing essentially

**Comment by justahuman1** [comment]
Can this issue now be closed given #17754 or is there remaining work? cc: @sajjad-moradi
